### PR TITLE
Bhangmeter re-write and other explosive goodness

### LINKED ISSUE #25104: Panning the map doesn't work in NanoUI minimaps like the Bhangmeter and the Central Atmospherics Computer
(WEB REPORT BY: burneddi REMOTE: 206.221.180.138:7777)
Panning the map by clicking and dragging does not appear to work on these consoles. It does work on the Advanced Security Cameras, because that's not real NanoUI.

### LINKED ISSUE #31453: The Bhangmeter (on roid, at least) is cursed and causes your game to lag into hell and eventually crash
(WEB REPORT BY: dakkanob REMOTE: 206.221.180.138:7777)
# Revision
81f4042e2a3ea8abea6a0cc816a5a0dc90e8e28c

# Description
title
# Steps to Reproduce
touch bhangmeter on roid
# What you Expected
frames
# What Actually Happened
features so bad they send your soul to space brazil

### LINKED ISSUE #33395: Bhangmeter crashes game
(WEB REPORT BY: knife28 REMOTE: 206.221.180.138:7777)
# Revision
620cb0e8cecdd5d50478272f973333cd53fce3aa

# Description
Not sure if its just on my end but opening the Bhangmeter caused my BYOND to go non-responsive and forced me to close and re-log. Same thing happened multiple times trying to use the console. 
# Steps to Reproduce

# What you Expected

# What Actually Happened

### PR DESCRIPTION
**Pull request by DeityLink**
Fixes #33395
Fixes #31453 
Fixes #25104

bhangmeter

The Bhangmeter has finally received the tender love and care it desperately needed.

image
image
meteor warning
image

imageimage
image
image

Additionally the `/datum/subsystem/obj/Initialize()` and `generateHoloMinimap()` procs have been set to "background=1" (which the game does automatically right now but this way it stops filling Dream Daemon's window with logs about it)

:cl:
* rscadd: The Bhangmeter now functions like a Station Holomap. Touching it displays a map of the station over which are displayed the areas affected by explosions. Recent explosions appear bright and opaque, then slowly fade away over 5 minutes until they remain at half transparency. Incoming meteor/object waves and the direction they come from are also indicated.
* rscadd: You can access the new Bhangmeter panel featuring a table of all recorded explosions by clicking the Database interface button visible on the holomap. Explosions are recorded from most recent to oldest, and per Z level, so if you want to see the Asteroid's explosions, you'll need to use a Bhangmeter on the Asteroid Z level.
* rscadd: Explosions caused by the Pipe Gun (aka Blast Canon) now spread in a cone shape. Just like with regular explosions, each affected tile now checks for things that could dampen the explosion between them and the epicenter.
* rscadd: The round-end scoreboard now gives the round's largest TTV explosion, with an extra mention if it reached maxcap.
* imageadd: Added additional in-hand sprites for TTV depending on which tanks are on them.
* imageadd: Improved the Pipe Gun (aka Blast Cannon)'s in-hand sprite, and added a variation for when it has a TTV equipped on it.
* imageadd: The Bhangmeter has a new sprite, finally differentiating it from the Detective's Forensic computer.
* bugfix: When the bhangmeter announces information, it does so by "coldly stating" which it hasn't done since at least 2016.
* rscdel: As they were mostly broken and unused, and extended greatly the duration of roundstart initialization, NanoUI minimap generation has been disabled. The only feature that still uses them is the Central Atmospherics Computer which will now display its markers over empty space when you display its Tracking Map. In return, we've shaved off 10-20 seconds of roundstart init.

### CODE REVIEW

**Comment by Eneocho** [comment]
Gonna miss pranking the newer players with the bhangmeter, and the ability to check for structures outside of Z1 with the minimap, but good stuff.

**Comment by DeityLink** [comment]
fixed conflicts

File: code/game/machinery/computer/bhangmeter.dm [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+				M.client.images -= watcher_maps["\ref[M]"]
+				qdel(watcher_maps["\ref[M]"])
+				watcher_maps["\ref[M]"] = image(holomap_datum.get_bhangmap())
+				var/image/I = watcher_maps["\ref[M]"]
+				I.loc = M.hud_used.holomap_obj
+				M.client.images |= watcher_maps["\ref[M]"]
+	else
+		stopWatching()
+
+/obj/machinery/computer/bhangmeter/say_quote(text)
```
Comment (Kurfursten) [inline code/game/machinery/computer/bhangmeter.dm:77]: objects can use ordinary say now, and machines should automatically beep as their sayverb

EDIT: Nevermind, I see from your description your intent was to override this sayverb

File: code/controllers/subsystem/objects.dm [hunk 2]
```diff
@@ -16,6 +16,7 @@ var/list/processing_objects = list()
 	NEW_SS_GLOBAL(SSobj)
 
 /datum/subsystem/obj/Initialize()
+	set background=1
```
Comment (Kurfursten) [inline code/controllers/subsystem/objects.dm:19]: what purpose is this ssobj init change serving?

File: code/game/machinery/computer/bhangmeter.dm [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+	return "coldly states, [text]"
+
+/obj/machinery/computer/bhangmeter/attack_hand(var/mob/user)
+	if(isliving(user) && anchored && !(stat & (FORCEDISABLE|NOPOWER|BROKEN)))
```
Comment (Kurfursten) [inline code/game/machinery/computer/bhangmeter.dm]: attack_ghost points here but this sanity check kicks out observers, but it also doesn't care if you are dead or a mouse

you can also invert the check with a return to avoid rightward creep

File: code/game/machinery/computer/bhangmeter.dm [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+	attack_hand(user)
+
+/obj/machinery/computer/bhangmeter/attack_animal(var/mob/user)
+	attack_hand(user)
+
+/obj/machinery/computer/bhangmeter/attack_ghost(var/mob/user)
+	attack_hand(user)
+
+/obj/machinery/computer/bhangmeter/interface_act(var/mob/i_user,var/action)
+	switch(action)
```
Comment (Kurfursten) [inline code/game/machinery/computer/bhangmeter.dm]: switch contains only one action...?

File: code/controllers/subsystem/objects.dm (hunk 2 above)
Comment (DeityLink) [reply code/controllers/subsystem/objects.dm:19]: This proc always gets sent to the background regardless on server init, causing a message log to pop in Dream Daemon (which makes sense given the amount of objects). This line just makes it so it does so right away and doesn't send the message. Mostly to unclog the boot up logs.
image

File: code/game/machinery/computer/bhangmeter.dm [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+/obj/machinery/computer/bhangmeter/proc/announce_explosion(var/datum/sensed_explosion/SE)
+	if(stat & (FORCEDISABLE|NOPOWER|BROKEN))
+		return
+	if (SE.z != original_zLevel)
+		return
+	if (last_announced_explosion && (world.time < (last_announcement + announcement_cooldown)))
+		var/new_score = SE.dev * 4 + SE.heavy * 2 + SE.light
+		var/old_score = last_announced_explosion.dev * 4 + last_announced_explosion.heavy * 2 + last_announced_explosion.light
+		if (old_score >= new_score)
+			return//We add a delay between announcements, unless the new explosion is larger than the last one.
```
Comment (Kurfursten) [inline code/game/machinery/computer/bhangmeter.dm:220]: looks like you might be exiting before an intended potential speech behavior below, consider re-ordering

File: code/modules/admin/verbs/mapping.dm [hunk 6]
```diff
@@ -161,7 +161,7 @@ var/intercom_range_display_status = 0
 	src.verbs += /client/proc/Test_ZAS_Connection
 	src.verbs += /client/proc/SDQL2_query
 	src.verbs += /client/proc/check_sim_unsim
-	src.verbs += /proc/generateMiniMaps
+	//src.verbs += /proc/generateMiniMaps
```
Comment (Kurfursten) [inline code/modules/admin/verbs/mapping.dm]: ?

File: code/game/machinery/computer/bhangmeter.dm (hunk 1 above)
Comment (DeityLink) [reply code/game/machinery/computer/bhangmeter.dm:77]: what do you mean in that context? Do you want the computer to "beep" instead of "coldly states" ?

File: code/modules/projectiles/guns/projectile/constructable/blastcannon.dm [hunk 7]
```diff
@@ -17,10 +18,14 @@
 	conventional_firearm = 0
 	silenced = 1
 	var/obj/item/device/transfer_valve/bomb = null
+	var/obj/item/toy/bomb/toybomb = null
 	var/datum/gas_mixture/bomb_air_contents_1 = null
 	var/datum/gas_mixture/bomb_air_contents_2 = null
 	var/ignorecap = 0
+	var/overcap = 0
 	var/bomb_appearance = null
+	var/widening_rate = 5
```
Comment (Kurfursten) [inline code/modules/projectiles/guns/projectile/constructable/blastcannon.dm:27]: this whole widening cone revamp to the blast cannon is such a significant change that it really should be its own PR for atomicity, but no one seems to be objecting to it so maybe it's fine

Perhaps people just don't see blast cannons enough these days but they were very controversial once upon a time, and I feel like people might have missed this change in the large gamut of other changes.

File: code/game/machinery/computer/bhangmeter.dm (hunk 5 above)
Comment (DeityLink) [reply code/game/machinery/computer/bhangmeter.dm:220]: well no, the point of that return IS to exit the proc without making any announcement.

If more than one explosion occurs within 5 seconds, only the first one gets announced, unless the second one is larger.

File: code/modules/admin/verbs/mapping.dm (hunk 6 above)
Comment (DeityLink) [reply code/modules/admin/verbs/mapping.dm]: you're right, might as well remove it now.

File: code/modules/projectiles/guns/projectile/constructable/blastcannon.dm (hunk 7 above)
Comment (DeityLink) [reply code/modules/projectiles/guns/projectile/constructable/blastcannon.dm:27]: People were very joyous about it when I displayed the blast canon changes on Discord. If it proves to be a bad idea in hindsight, simply editing that var to something like 999 will cause the blast canon to fire in a thin line again.

Comment (DeityLink) [reply code/modules/projectiles/guns/projectile/constructable/blastcannon.dm:27]: Also worth noting, they might have been controversial due to ignoring the explosion blocking effect of walls and other obstacles, which the new code takes into account.
